File: engineering/LeetCode/543.Diameter-of-Binary-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
https://leetcode.com/problems/diameter-of-binary-tree/description/

Time complexity: O(n), n = len(root)
Space complexity: O(1)
"""
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def find_max_path(node: Optional[TreeNode]) -> int:
            nonlocal res
            
            if not node: return 0
            left, right = find_max_path(node.left), find_max_path(node.right)
            res = max(res, left + right)
            return max(left, right) + 1

        find_max_path(root)
        return res

File: engineering/LeetCode/543.diameter-of-binary-tree.ts
/*
 * @lc app=leetcode id=543 lang=typescript
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    if (root === null) { return 0; }

    const left = root.left ? diameterOfBinaryTree(root.left) : 0;
    const right = root.right ? diameterOfBinaryTree(root.right) : 0;

    return (left + right) > 0 ? 1 : 0 + left + right;
};
// @lc code=end

